src/domain/Processors/VehicleProcessor.ts

```diff
@@ -0,0 +1,533 @@
+   */
+  private async createAndArchiveTechRecord(
+    updatedVehicle: T,
+    msUserDetails: IMsUserDetails,
+    oldStatusCode: STATUS | undefined
+  ): Promise<T> {
+    const { statusCode } = updatedVehicle.techRecord[0];
+
+    VehicleProcessor.checkStatusCodeValidity(statusCode, oldStatusCode);
+
+    updatedVehicle.techRecord[0] = this.validate(updatedVehicle, false);
+
+    try {
+      const techRecordWithAllStatuses = await this.getTechRecordList(
+        updatedVehicle.systemNumber
+      );
+
+      const errors = await this.validateVrmWithHistory(updatedVehicle, techRecordWithAllStatuses);
+      if(errors && errors.length) {
+        throw ErrorHandler.Error(400, errors);
+      }
+      const techRecToArchive = VehicleProcessor.getTechRecordToArchive(
+        techRecordWithAllStatuses,
+        oldStatusCode ? oldStatusCode : statusCode
+      );
+
+      // if status code has changed from provisional to current
+      this.updateCurrentStatusCode(
+        oldStatusCode,
+        statusCode,
+        techRecordWithAllStatuses,
+        msUserDetails
+      );
+
+      this.updateVehicleIdentifiers(
+        techRecordWithAllStatuses,
+        updatedVehicle
+      );
+
+      this.capitaliseGeneralVehicleAttributes(updatedVehicle);
+
+      const newRecord: TechRecord = cloneDeep(techRecToArchive);
+      mergeWith(
+        newRecord,
+        updatedVehicle.techRecord[0],
+        VehicleProcessor.arrayCustomizer
+      );
+      if (oldStatusCode) {
+        newRecord.statusCode = statusCode;
+      }
+      AuditDetailsHandler.setAuditDetails(
+        newRecord,
+        techRecToArchive,
+        msUserDetails
+      );
+      techRecToArchive.statusCode = STATUS.ARCHIVED;
+      this.mapFields(newRecord);
+      const { systemNumber, vin, primaryVrm } = techRecordWithAllStatuses;
+      newRecord.recordCompleteness = this.calculateRecordCompleteness({
+        systemNumber,
+        vin,
+        primaryVrm,
+        techRecord: [newRecord]
+      } as T);
+      techRecordWithAllStatuses.techRecord.push(newRecord);
+      return techRecordWithAllStatuses;
+    } catch (error) {
+      console.error(error);
+      throw ErrorHandler.Error(error.statusCode, error.body);
+    }
+  }
+
+  /**
+   * To validate common attributes for all vehicles and specific attributes based on vehicle type.
+   * @param newVehicle vehicle to validate
+   * @param isCreate true if it is a create request else false
+   */
+  private validate(newVehicle: T, isCreate: boolean): TechRecord {
+    let errors: string[] = [];
+    const isPrimaryVrmRequired= this.vehicle.techRecord[0].vehicleType !== VEHICLE_TYPE.TRL;
+    const {primaryVrm, secondaryVrms} = this.vehicle;
+    // primary & secondary vrms are required in case of create and optional in case of update
+    const validatePrimaryVrm = isCreate || primaryVrm;
+    const validateSecondaryVrms = isCreate || secondaryVrms;
+    errors =  errors.concat(
```

Comment:
I didn't finish but started to think about refactoring this if my understanding is correct
we return an array or errors if some attributes (vrms) are missing so rather than having a ternary operator in the concat perhaps we could use something like that as a first step

```
// We could also partially apply some arguments so the validator becomes generic for primary VRM & secondary vrm later

const primaryVrmValidator = (primaryVrm, isRequired) => {
  const validationResult = validatePrimaryVrm(primaryVrm);
  if((isRequired && !primaryVrm) || validationResult) return ['PRIMARY_INVALID'];
  else []
};
```

and pipe the concat
```
errors = errors
  .concat(primaryVrmValidator(primaryVrm, isPrimaryVrmRequired))
  .concat()
```